### PR TITLE
Improve Native.Call performance and add Native.Call overloads with explicit native args

### PR DESCRIPTION
**Pull request by kagikn**
Why I created PR like this? Because I think heap allocation should not happen so much for native calls in common cases.

Thy these regular expressions with natives.h generated from [alloc8or's Native DB](https://alloc8or.re/gta5/nativedb/). Line counts are for natives.h for b2372.
no args: `>\(0x[0-9A-F]{1,16}\)` (Line count: 1043)
one arg: `>\(0x[0-9A-F]{1,16}, [A-Za-z0-9]{1,32}\)` (Line count: 2242)
two args: `>\(0x[0-9A-F]{1,16}, [A-Za-z0-9]{1,32}, [A-Za-z0-9]{1,32}\)` (Line count: 1508)
three args: `>\(0x[0-9A-F]{1,16}, [A-Za-z0-9]{1,32}, [A-Za-z0-9]{1,32}, [A-Za-z0-9]{1,32}\)` (Line count: 546)
four args: `>\(0x[0-9A-F]{1,16}, [A-Za-z0-9]{1,32}, [A-Za-z0-9]{1,32}, [A-Za-z0-9]{1,32}, [A-Za-z0-9]{1,32}\)` (Line count: 357)
natives with between no args to four args: (1043 + 2242 + 1508 + 546 + 357) = 5696
total natives: `NATIVE_INLINE static` (Line count: 6350)

So I can say this PR will add `Native.Call` overloads for about 90% natives to avoid heap allocation for the param array. We could define `Native.Call` overloads with 5 to 16 args so SHVDN can avoid `params` array allocations for almost all natives that some scripts call frequently (like 60 times on 1 second), such as `START_EXPENSIVE_SYNCHRONOUS_SHAPE_TEST_LOS_PROBE`, `DRAW_MARKER`, `DRAW_SPRITE`, and `DRAW_SPOT_LIGHT`.
If `InputArgument` was defined as struct, things would be better but it's too late.
_I would create 33 overloads for `Native.Call` (with no params to 32 params or at least 16 ones) if I was the owner of this repo or if I create .NET 5 (.NET Core in other words) variant of SHVDN, btw._

### CODE REVIEW

File: source/scripting_v3/GTA.Native/Native.cs [hunk 1]
```diff
@@ -358,56 +360,254 @@ public T GetResult<T>()
...
+
+				argPtr[0] = argument0._data;
+				argPtr[1] = argument1._data;
+				argPtr[2] = argument2._data;
+				argPtr[3] = argument3._data;
+
+				return CallInternalNoParamArray<T>(hash, argPtr, 4);
+			}
+		}
+		static unsafe T CallInternalNoParamArray<T>(Hash hash, ulong* argPtr, int argCount)
```
Comment (crosire) [inline source/scripting_v3/GTA.Native/Native.cs:477]: Not sure this function really helps readability. It's just two lines, so may as well duplicate those everywhere.

**Review by crosire — COMMENTED** [review]
Bit surprised this just works and ones doesn't run into problems with multiple overloads being defined for the same thing (params overload vs multiple parameters), but I guess normal parameters take precedence.

**Comment by kagikn** [comment]
Normal parameters take precedence over params overload unless you pass an array explicitly when compiling, yes. Compiled scripts still use params one no matter how many parameters are, though.

**Comment by kagikn** [comment]
Do you mind adding overloads with 5 to 16 `NativeArgument`s? There are many useful natives that take 5 to 16 params as I said before (There are few natives that take 17 to 32 params).

**Comment by crosire** [comment]
I don't really mind. At this point the difference between 5 or 16 overloads doesn't really matter, the file is getting big anyway. And can just collpase them when scrolling through.

**Comment by kagikn** [comment]
I don’t think this PR has any compatibility problems. If there were other issues, it would be how to format `Native.Call` with large set of parameters.

**Pull request merged**